src/mlnet/Telemetry/Telemetry.cs

```diff
@@ -1,97 +1,146 @@
 
-namespace Microsoft.DotNet.Cli.Telemetry
+namespace Microsoft.ML.CLI.Telemetry
 {
-    public class Telemetry : ITelemetry
+    /// <summary>
+    /// Houses CLI telemetry collection and utility methods.
+    /// </summary>
+    internal static class Telemetry
     {
-        private TelemetryClient _client;
-        private Dictionary<string, string> _commonProperties;
-        private Task _trackEventTask;
+        private static TelemetryClient _client;
+        private static Dictionary<string, string> _commonProperties;
+        private static bool _enabled;
+        private static Task _initializationTask;
 
         private const string InstrumentationKey = "c059917c-818d-489a-bfcb-351eaab73f2a";
         private const string MlTelemetryOptout = "MLDOTNET_CLI_TELEMETRY_OPTOUT";
         private const string MachineId = "MachineId";
 
-        public bool Enabled { get; }
-
-        public Telemetry()
+        public static void Initialize()
         {
-             _commonProperties = new Dictionary<string, string>();
             var optedOut = Env.GetEnvironmentVariableAsBool(MlTelemetryOptout, false);
-
-            Enabled = !optedOut;
-
-            if (!Enabled)
+            _enabled = !optedOut;
+            if (!_enabled)
             {
                 return;
             }
 
-            //initialize in task to offload to parallel thread
-            _trackEventTask = Task.Factory.StartNew(() => InitializeTelemetry());
+            // Initialize in task to offload to parallel thread
+            _initializationTask = Task.Factory.StartNew(() => InitializeTask());
         }
 
-        public void TrackEvent(
+        /// <summary>
+        /// Send telemetry event.
+        /// </summary>
+        public static void TrackEvent(
             string eventName,
             IDictionary<string, string> properties,
-            IDictionary<string, double> measurements)
+            TimeSpan? duration = null,
+            Exception ex = null)
         {
-            if (!Enabled)
+            if (!_enabled)
             {
                 return;
             }
-
-            //continue task in existing parallel thread
-            _trackEventTask = _trackEventTask.ContinueWith(
-                x => TrackEventTask(eventName, properties, measurements)
-            );
+            _initializationTask.ContinueWith(x => TrackEventTask(eventName, properties, duration, ex));
         }
 
-        public void ThreadBlockingTrackEvent(string eventName, IDictionary<string, string> properties, IDictionary<string, double> measurements)
+        /// <summary>
+        /// Flush outstanding telemetry, and wait for the specified timeout for this to complete.
+        /// </summary>
+        public static void Flush(TimeSpan timeout)
         {
-            if (!Enabled)
+            if (!_enabled || _client == null)
             {
                 return;
             }
+            Task.Run(() => _client.Flush()).Wait(timeout);
```

Comment:
A bit if a nit but this will block the thread.  Probably not relevant in the way it is used currently since it's only when everything else is done.  But can become an issue in the future.   It's possible to do it async:  await Task.WhenAny(task, Task.Delay(timeout))